File: 2nd year/MPP/project-mpp/java-project/TriathlonPersistence/src/main/java/triathlon/persistence/dbRepository/DBGameRepository.java
package triathlon.persistence.dbRepository;

import triathlon.persistence.validators.ValidationException;
import triathlon.persistence.RepositoryException;
import triathlon.persistence.GameRepository;
import triathlon.model.Game;
import triathlon.persistence.validators.Validator;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

public class DBGameRepository implements GameRepository {
    private final Validator<Game> validator;
    private final Connection connection;

    public DBGameRepository(Properties props, Validator<Game> validator) {
        this.connection = new JDBCManager(props).getConnection();
        this.validator = validator;
    }

    @Override
    public Game findOne(Long id) {
        final String sql_query = "select * from Game where id = ?";

        try (PreparedStatement statement = connection.prepareStatement(sql_query)) {
            statement.setLong(1, id);

            ResultSet resultSet = statement.executeQuery();
            if (resultSet.next()) {
                return new Game(
                        resultSet.getLong("id"),
                        resultSet.getString("name")
                );
            }
        } catch (SQLException e) {
            System.err.println("Error DB: " + e);
        }

        return null;
    }

    @Override
    public Iterable<Game> findAll() {
        final String sql_query = "select * from Game";

        try (PreparedStatement statement = connection.prepareStatement(sql_query)) {
            List<Game> games = new ArrayList<>();
            ResultSet resultSet = statement.executeQuery();
            while (resultSet.next()) {
                games.add(new Game(
                        resultSet.getLong("id"),
                        resultSet.getString("name")
                ));
            }
            return games;
        } catch (SQLException e) {
            System.err.println("Error DB: " + e);
        }

        return null;
    }

    @Override
    public void save(Game game) {
        final String sql_query = "insert into Game (name) values (?)";
        try {
            validator.validate(game);
        } catch (ValidationException ex) {
            throw new RepositoryException(ex.getMessage());
        }

        try (PreparedStatement statement = connection.prepareStatement(sql_query)) {
            statement.setString(1, game.getName());
            int updates = statement.executeUpdate();
        } catch (SQLException e) {
            System.err.println("Error DB: " + e);
        }
    }

    @Override
    public void update(Game game) {
        final String sql_query = "update Game set name = ? where id = ?";
        try {
            validator.validate(game);
        } catch (ValidationException ex) {
            throw new RepositoryException(ex.getMessage());
        }

        try (PreparedStatement statement = connection.prepareStatement(sql_query)) {
            statement.setString(1, game.getName());
            statement.setLong(2, game.getId());
            int updates = statement.executeUpdate();
        } catch (SQLException e) {
            System.err.println("Error DB: " + e);
        }
    }

    @Override
    public void delete(Long id) {
        final String sql_query = "delete from Game where id = ?";

        try (PreparedStatement statement = connection.prepareStatement(sql_query)) {
            statement.setLong(1, id);

            int updates = statement.executeUpdate();
        } catch (SQLException e) {
            System.err.println("Error DB: " + e);
        }
    }
}

File: 2nd year/MPP/project-mpp/csharp-project/persistence/dbRepository/DbGameRepository.cs
using System;
using System.Collections.Generic;
using log4net;
using Microsoft.Data.Sqlite;
using model;
using persistence.validators;

namespace persistence.dbRepository
{
    public class DbGameRepository : IGameRepository
    {
        private static readonly ILog Log = LogManager.GetLogger(typeof(DbGameRepository));
        private readonly SqliteConnection _connection = JdbcManager.GetConnection();
        private readonly IValidator<Game> _validator;

        public DbGameRepository(IValidator<Game> validator)
        {
            this._validator = validator;
        }

        public Game FindOne(long id)
        {
            Log.Info($"Entered with id = {id}");
            try
            {
                const string commandText = "select * from Game where id = @id";
                using var command = new SqliteCommand(commandText, _connection);
                Log.Debug(commandText);

                command.Parameters.AddWithValue("id", id);

                using var reader = command.ExecuteReader();
                if (reader.Read())
                {
                    return new Game
                    {
                        Id = reader.GetInt64(0),
                        Name = reader.GetString(1)
                    };
                }
            }
            catch (Exception ex)
            {
                Log.Error(ex.Message);
                Console.Write(ex.Message);
            }

            return null;
        }

        public IEnumerable<Game> FindAll()
        {
            Log.Info("");
            const string commandText = "select * from Game";
            IList<Game> games = new List<Game>();

            try
            {
                using var command = new SqliteCommand(commandText, _connection);
                Log.Debug(commandText);

                using var reader = command.ExecuteReader();
                while (reader.Read())
                {
                    games.Add(new Game
                    {
                        Id = reader.GetInt64(0),
                        Name = reader.GetString(1)
                    });
                }
            }
            catch (Exception ex)
            {
                Log.Error(ex.Message);
                Console.Write(ex.Message);
            }

            Log.Info(games);
            return games;
        }

        public void Save(Game game)
        {
            Log.Info($"Entered with entity = {game}");
            const string commandText = "insert into Game (id, name) values (@id, @name)";
            _validator.Validate(game);
            try
            {
                using var command = new SqliteCommand(commandText, _connection);
                Log.Debug(commandText);

                command.Parameters.AddWithValue("id", game.Id);
                command.Parameters.AddWithValue("name", game.Name);

                command.ExecuteNonQuery();
            }
            catch (Exception ex)
            {
                Log.Error(ex.Message);
                Console.Write(ex.Message);
            }
        }

        public void Update(Game game)
        {
            Log.Info($"Entered with entity = {game}");
            const string commandText = "update Game set name = @name where id = @id";
            _validator.Validate(game);
            try
            {
                using var command = new SqliteCommand(commandText, _connection);
                Log.Debug(commandText);

                command.Parameters.AddWithValue("id", game.Id);
                command.Parameters.AddWithValue("name", game.Name);

                command.ExecuteNonQuery();
            }
            catch (Exception ex)
            {
                Log.Error(ex.Message);
                Console.Write(ex.Message);
            }
        }

        public void Delete(long id)
        {
            Log.Info($"Entered with id = {id}");
            const string commandText = "delete from Game where id = @id";
            try
            {
                using var command = new SqliteCommand(commandText, _connection);
                Log.Debug(commandText);

                command.Parameters.AddWithValue("id", id);

                command.ExecuteNonQuery();
            }
            catch (Exception ex)
            {
                Log.Error(ex.Message);
                Console.Write(ex.Message);
            }
        }
    }
}
